lib/puppet/util/connection.rb

```diff
@@ -22,15 +22,17 @@ module Connection
     # @return [String] the name of the server for use in the request
     def self.determine_server(setting)
       if setting && setting != :server && Puppet.settings.set_by_config?(setting)
+        Puppet.debug _("Selected server from the %{setting} setting: %{server}") % {setting: setting, server: Puppet.settings[setting]}
         Puppet[setting]
       else
         server = Puppet.lookup(:server) do
           if primary_server = Puppet.settings[:server_list][0]
-            Puppet.debug "Dynamically-bound server lookup failed; using first entry"
```

Comment:
@joshcooper Charlie's commit to change `Puppet.debug` to `debug_once` https://github.com/puppetlabs/puppet/commit/179f3e8e5d0cc059ccc4e77a96e1d828b716015d is in a later branch